File: src/bin/tui/ansi.rs
//! ANSI SGR → `ratatui` spans.
//!
//! `mudlib/lib/color.lua` emits a known, small set: `ESC[Nm` for the 16 colours,
//! the bright pair, the styles and reset, plus `ESC[38;5;Nm` / `ESC[48;5;Nm` for
//! its xterm-256 aliases. Everything here covers that, plus 24-bit colour and
//! the "turn it off again" codes for completeness, and ignores any other CSI
//! sequence rather than printing it.
//!
//! The decoder is a byte state machine because output arrives in arbitrary TCP
//! chunks: an escape sequence, a UTF-8 character, or a line can be split across
//! two reads. Style carries across line breaks, as it does on a real terminal.

use ratatui::style::{Color, Modifier, Style};
use ratatui::text::{Line, Span};

#[derive(Debug, Clone, Copy, PartialEq)]
enum State {
    Normal,
    Escape,
    Csi,
}

#[derive(Default)]
pub struct AnsiDecoder {
    state: State,
    /// Parameter bytes of the CSI sequence currently being read.
    params: Vec<u8>,
    /// Style in force, carried across lines.
    style: Style,
    /// Spans completed on the line being built.
    spans: Vec<Span<'static>>,
    /// Bytes of the span currently being built, decoded when the span closes.
    text: Vec<u8>,
}

impl Default for State {
    fn default() -> Self {
        State::Normal
    }
}

impl AnsiDecoder {
    pub fn new() -> Self {
        Self::default()
    }

    /// Feed a chunk of game output. Returns every line completed by it; any
    /// trailing partial line stays buffered and is readable via [`partial`].
    ///
    /// [`partial`]: AnsiDecoder::partial
    pub fn feed(&mut self, bytes: &[u8]) -> Vec<Line<'static>> {
        let mut lines = Vec::new();
        for &b in bytes {
            match self.state {
                State::Normal => match b {
                    0x1b => self.state = State::Escape,
                    b'\n' => lines.push(self.take_line()),
                    // The mudlib sends CRLF; a bare CR is a carriage return we
                    // have no use for either, since we rebuild the line anyway.
                    b'\r' => {}
                    // Telnet NUL padding after CR, and the BEL we do not ring.
                    0x00 | 0x07 => {}
                    _ => self.text.push(b),
                },
                State::Escape => {
                    if b == b'[' {
                        self.params.clear();
                        self.state = State::Csi;
                    } else {
                        // Not a CSI — ESC ) 0 and friends. Drop the introducer
                        // and treat the next byte as ordinary text.
                        self.state = State::Normal;
                    }
                }
                State::Csi => {
                    // A CSI ends at the first byte in 0x40..=0x7e; everything
                    // before it is parameter or intermediate bytes.
                    if (0x40..=0x7e).contains(&b) {
                        if b == b'm' {
                            self.apply_sgr();
                        }
                        self.state = State::Normal;
                    } else {
                        self.params.push(b);
                    }
                }
            }
        }
        lines
    }

    /// The partial line currently buffered — the prompt, which the driver sends
    /// without a trailing newline. Returns `None` when nothing is pending.
    pub fn partial(&self) -> Option<Line<'static>> {
        if self.text.is_empty() && self.spans.is_empty() {
            return None;
        }
        let mut spans = self.spans.clone();
        if !self.text.is_empty() {
            spans.push(Span::styled(
                String::from_utf8_lossy(&self.text).into_owned(),
                self.style,
            ));
        }
        Some(Line::from(spans))
    }

    fn close_span(&mut self) {
        if !self.text.is_empty() {
            let text = String::from_utf8_lossy(&self.text).into_owned();
            self.spans.push(Span::styled(text, self.style));
            self.text.clear();
        }
    }

    fn take_line(&mut self) -> Line<'static> {
        self.close_span();
        Line::from(std::mem::take(&mut self.spans))
    }

    fn apply_sgr(&mut self) {
        // An empty parameter list — a bare `ESC[m` — means reset, same as `0`.
        let raw = String::from_utf8_lossy(&self.params);
        let codes: Vec<u16> = if raw.is_empty() {
            vec![0]
        } else {
            raw.split(';')
                .map(|p| p.trim().parse::<u16>().unwrap_or(0))
                .collect()
        };

        let before = self.style;
        let mut i = 0;
        while i < codes.len() {
            match codes[i] {
                0 => self.style = Style::default(),
                1 => self.style = self.style.add_modifier(Modifier::BOLD),
                2 => self.style = self.style.add_modifier(Modifier::DIM),
                3 => self.style = self.style.add_modifier(Modifier::ITALIC),
                4 => self.style = self.style.add_modifier(Modifier::UNDERLINED),
                5 => self.style = self.style.add_modifier(Modifier::SLOW_BLINK),
                7 => self.style = self.style.add_modifier(Modifier::REVERSED),
                9 => self.style = self.style.add_modifier(Modifier::CROSSED_OUT),
                22 => {
                    self.style = self
                        .style
                        .remove_modifier(Modifier::BOLD | Modifier::DIM)
                }
                23 => self.style = self.style.remove_modifier(Modifier::ITALIC),
                24 => self.style = self.style.remove_modifier(Modifier::UNDERLINED),
                25 => self.style = self.style.remove_modifier(Modifier::SLOW_BLINK),
                27 => self.style = self.style.remove_modifier(Modifier::REVERSED),
                29 => self.style = self.style.remove_modifier(Modifier::CROSSED_OUT),
                c @ 30..=37 => self.style = self.style.fg(basic(c - 30)),
                38 => {
                    if let Some((color, used)) = extended(&codes[i..]) {
                        self.style = self.style.fg(color);
                        i += used - 1;
                    }
                }
                39 => self.style = self.style.fg(Color::Reset),
                c @ 40..=47 => self.style = self.style.bg(basic(c - 40)),
                48 => {
                    if let Some((color, used)) = extended(&codes[i..]) {
                        self.style = self.style.bg(color);
                        i += used - 1;
                    }
                }
                49 => self.style = self.style.bg(Color::Reset),
                c @ 90..=97 => self.style = self.style.fg(bright(c - 90)),
                c @ 100..=107 => self.style = self.style.bg(bright(c - 100)),
                // Anything else is a code the mudlib does not emit; ignoring it
                // is better than rendering the escape as text.
                _ => {}
            }
            i += 1;
        }

        // Only break the span if the style actually moved. A no-op `ESC[m` in
        // the middle of a word should not fragment it.
        if self.style != before {
            let after = self.style;
            self.style = before;
            self.close_span();
            self.style = after;
        }
    }
}

/// `38`/`48` take a sub-form: `;5;N` for indexed, `;2;R;G;B` for 24-bit.
/// Returns the colour and how many codes it consumed.
fn extended(codes: &[u16]) -> Option<(Color, usize)> {
    match codes.get(1)? {
        5 => Some((Color::Indexed(*codes.get(2)? as u8), 3)),
        2 => {
            let r = *codes.get(2)? as u8;
            let g = *codes.get(3)? as u8;
            let b = *codes.get(4)? as u8;
            Some((Color::Rgb(r, g, b), 5))
        }
        _ => None,
    }
}

fn basic(n: u16) -> Color {
    match n {
        0 => Color::Black,
        1 => Color::Red,
        2 => Color::Green,
        3 => Color::Yellow,
        4 => Color::Blue,
        5 => Color::Magenta,
        6 => Color::Cyan,
        _ => Color::Gray,
    }
}

fn bright(n: u16) -> Color {
    match n {
        0 => Color::DarkGray,
        1 => Color::LightRed,
        2 => Color::LightGreen,
        3 => Color::LightYellow,
        4 => Color::LightBlue,
        5 => Color::LightMagenta,
        6 => Color::LightCyan,
        _ => Color::White,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Flatten a line to `(text, style)` pairs for assertions.
    fn parts(line: &Line<'static>) -> Vec<(String, Style)> {
        line.spans
            .iter()
            .map(|s| (s.content.to_string(), s.style))
            .collect()
    }

    #[test]
    fn plain_text_splits_on_newlines_and_drops_cr() {
        let mut d = AnsiDecoder::new();
        let lines = d.feed(b"one\r\ntwo\r\n");
        assert_eq!(lines.len(), 2);
        assert_eq!(parts(&lines[0])[0].0, "one");
        assert_eq!(parts(&lines[1])[0].0, "two");
        assert!(d.partial().is_none());
    }

    #[test]
    fn a_trailing_partial_line_is_the_prompt() {
        let mut d = AnsiDecoder::new();
        let lines = d.feed(b"scrolled\r\n42h 20m> ");
        assert_eq!(lines.len(), 1);
        let prompt = d.partial().expect("prompt should be buffered");
        assert_eq!(parts(&prompt)[0].0, "42h 20m> ");
    }

    #[test]
    fn colour_codes_become_styled_spans() {
        let mut d = AnsiDecoder::new();
        // What `{red}danger{/} ok` colorizes to.
        let lines = d.feed(b"\x1b[31mdanger\x1b[0m ok\r\n");
        let p = parts(&lines[0]);
        assert_eq!(p.len(), 2);
        assert_eq!(p[0].0, "danger");
        assert_eq!(p[0].1.fg, Some(Color::Red));
        assert_eq!(p[1].0, " ok");
        assert_eq!(p[1].1.fg, None);
    }

    #[test]
    fn xterm_256_foreground_and_background() {
        let mut d = AnsiDecoder::new();
        // `{orange}` is 208; `{bg:17}` is the midnight background.
        let lines = d.feed(b"\x1b[38;5;208mA\x1b[48;5;17mB\r\n");
        let p = parts(&lines[0]);
        assert_eq!(p[0].1.fg, Some(Color::Indexed(208)));
        assert_eq!(p[1].1.fg, Some(Color::Indexed(208)));
        assert_eq!(p[1].1.bg, Some(Color::Indexed(17)));
    }

    #[test]
    fn truecolor_is_understood_even_though_the_mudlib_does_not_emit_it() {
        let mut d = AnsiDecoder::new();
        let lines = d.feed(b"\x1b[38;2;10;20;30mx\r\n");
        assert_eq!(parts(&lines[0])[0].1.fg, Some(Color::Rgb(10, 20, 30)));
    }

    #[test]
    fn modifiers_accumulate_and_clear() {
        let mut d = AnsiDecoder::new();
        let lines = d.feed(b"\x1b[1m\x1b[4mboth\x1b[24monly bold\r\n");
        let p = parts(&lines[0]);
        assert!(p[0].1.add_modifier.contains(Modifier::BOLD));
        assert!(p[0].1.add_modifier.contains(Modifier::UNDERLINED));
        assert!(p[1].1.add_modifier.contains(Modifier::BOLD));
        assert!(!p[1].1.add_modifier.contains(Modifier::UNDERLINED));
    }

    #[test]
    fn a_sequence_split_across_reads_still_parses() {
        // The failure this guards: an escape straddling a TCP chunk boundary
        // rendering as literal "[31m" in the game pane.
        let mut d = AnsiDecoder::new();
        assert!(d.feed(b"\x1b[3").is_empty());
        assert!(d.feed(b"1mred").is_empty());
        let lines = d.feed(b"\r\n");
        let p = parts(&lines[0]);
        assert_eq!(p[0].0, "red");
        assert_eq!(p[0].1.fg, Some(Color::Red));
    }

    #[test]
    fn a_utf8_character_split_across_reads_survives() {
        let mut d = AnsiDecoder::new();
        // The box-drawing the trace output uses, cut mid-codepoint.
        let full = "──".as_bytes();
        assert!(d.feed(&full[..3]).is_empty());
        let lines = d.feed(&[&full[3..], b"\r\n" as &[u8]].concat());
        assert_eq!(parts(&lines[0])[0].0, "──");
    }

    #[test]
    fn style_carries_across_a_line_break() {
        let mut d = AnsiDecoder::new();
        let lines = d.feed(b"\x1b[32mgreen\r\nstill green\r\n");
        assert_eq!(parts(&lines[0])[0].1.fg, Some(Color::Green));
        assert_eq!(parts(&lines[1])[0].1.fg, Some(Color::Green));
    }

    #[test]
    fn non_sgr_csi_sequences_are_swallowed_not_printed() {
        let mut d = AnsiDecoder::new();
        let lines = d.feed(b"a\x1b[2Jb\r\n");
        assert_eq!(parts(&lines[0])[0].0, "ab");
    }

    #[test]
    fn a_no_op_reset_does_not_fragment_a_word() {
        let mut d = AnsiDecoder::new();
        let lines = d.feed(b"unbro\x1b[0mken\r\n");
        assert_eq!(parts(&lines[0]).len(), 1);
    }
}

File: debug-client/src/lib/ansi.js
// ANSI SGR → styled spans.
//
// A port of `src/bin/tui/ansi.rs`. `mudlib/lib/color.lua` emits a known, small
// set: `ESC[Nm` for the 16 colours, the bright pair, the styles and reset, plus
// `ESC[38;5;Nm` / `ESC[48;5;Nm` for its xterm-256 aliases. Everything here
// covers that, plus 24-bit colour and the "turn it off again" codes for
// completeness, and ignores any other CSI sequence rather than printing it.
//
// The decoder is a state machine because output arrives in arbitrary chunks: an
// escape sequence or a line can be split across two frames. Style carries
// across line breaks, as it does on a real terminal.

const NORMAL = 0
const ESCAPE = 1
const CSI = 2

/// xterm's first 16, the palette every terminal disagrees about slightly. These
/// are the values the driver's own client uses.
const BASIC = [
  '#000000', '#cd0000', '#00cd00', '#cdcd00', '#0000ee', '#cd00cd', '#00cdcd', '#e5e5e5',
]
const BRIGHT = [
  '#7f7f7f', '#ff0000', '#00ff00', '#ffff00', '#5c5cff', '#ff00ff', '#00ffff', '#ffffff',
]
const CUBE = [0, 95, 135, 175, 215, 255]

const hex2 = (n) => n.toString(16).padStart(2, '0')

/// xterm-256 index → CSS colour.
export function indexed(n) {
  if (n < 8) return BASIC[n]
  if (n < 16) return BRIGHT[n - 8]
  if (n < 232) {
    const i = n - 16
    return `#${hex2(CUBE[Math.floor(i / 36) % 6])}${hex2(CUBE[Math.floor(i / 6) % 6])}${hex2(CUBE[i % 6])}`
  }
  const g = 8 + (n - 232) * 10
  return `#${hex2(g)}${hex2(g)}${hex2(g)}`
}

const EMPTY = Object.freeze({
  fg: null,
  bg: null,
  bold: false,
  dim: false,
  italic: false,
  underline: false,
  blink: false,
  reverse: false,
  strike: false,
})

function sameStyle(a, b) {
  for (const k of Object.keys(EMPTY)) if (a[k] !== b[k]) return false
  return true
}

export class AnsiDecoder {
  constructor() {
    this.state = NORMAL
    this.params = ''
    this.style = { ...EMPTY }
    this.spans = []
    this.text = ''
  }

  /// Feed a chunk of game output. Returns every line completed by it; any
  /// trailing partial line stays buffered and is readable via `partial()`.
  feed(chunk) {
    const lines = []
    for (const ch of chunk) {
      switch (this.state) {
        case NORMAL:
          if (ch === '\x1b') this.state = ESCAPE
          else if (ch === '\n') lines.push(this.#takeLine())
          // The mudlib sends CRLF; a bare CR is a carriage return we have no
          // use for either, since we rebuild the line anyway.
          else if (ch === '\r') { /* drop */ }
          // Telnet NUL padding after CR, and the BEL we do not ring.
          else if (ch === '\x00' || ch === '\x07') { /* drop */ }
          else this.text += ch
          break

        case ESCAPE:
          if (ch === '[') {
            this.params = ''
            this.state = CSI
          } else {
            // Not a CSI — ESC ) 0 and friends. Drop the introducer and treat
            // the next byte as ordinary text.
            this.state = NORMAL
          }
          break

        case CSI: {
          // A CSI ends at the first byte in 0x40..=0x7e; everything before it
          // is parameter or intermediate bytes.
          const code = ch.charCodeAt(0)
          if (code >= 0x40 && code <= 0x7e) {
            if (ch === 'm') this.#applySgr()
            this.state = NORMAL
          } else {
            this.params += ch
          }
          break
        }
      }
    }
    return lines
  }

  /// The partial line currently buffered — the prompt, which the driver sends
  /// without a trailing newline. Returns `null` when nothing is pending.
  partial() {
    if (this.text === '' && this.spans.length === 0) return null
    const spans = this.spans.slice()
    if (this.text !== '') spans.push({ text: this.text, style: { ...this.style } })
    return spans
  }

  #closeSpan() {
    if (this.text !== '') {
      this.spans.push({ text: this.text, style: { ...this.style } })
      this.text = ''
    }
  }

  #takeLine() {
    this.#closeSpan()
    const line = this.spans
    this.spans = []
    return line
  }

  #applySgr() {
    // An empty parameter list — a bare `ESC[m` — means reset, same as `0`.
    const codes =
      this.params === ''
        ? [0]
        : this.params.split(';').map((p) => {
            const n = Number.parseInt(p.trim(), 10)
            return Number.isFinite(n) ? n : 0
          })

    const before = { ...this.style }
    const s = this.style
    for (let i = 0; i < codes.length; i++) {
      const c = codes[i]
      if (c === 0) Object.assign(s, EMPTY)
      else if (c === 1) s.bold = true
      else if (c === 2) s.dim = true
      else if (c === 3) s.italic = true
      else if (c === 4) s.underline = true
      else if (c === 5) s.blink = true
      else if (c === 7) s.reverse = true
      else if (c === 9) s.strike = true
      else if (c === 22) { s.bold = false; s.dim = false }
      else if (c === 23) s.italic = false
      else if (c === 24) s.underline = false
      else if (c === 25) s.blink = false
      else if (c === 27) s.reverse = false
      else if (c === 29) s.strike = false
      else if (c >= 30 && c <= 37) s.fg = BASIC[c - 30]
      else if (c === 38) {
        const ext = extended(codes, i)
        if (ext) { s.fg = ext.color; i += ext.used - 1 }
      } else if (c === 39) s.fg = null
      else if (c >= 40 && c <= 47) s.bg = BASIC[c - 40]
      else if (c === 48) {
        const ext = extended(codes, i)
        if (ext) { s.bg = ext.color; i += ext.used - 1 }
      } else if (c === 49) s.bg = null
      else if (c >= 90 && c <= 97) s.fg = BRIGHT[c - 90]
      else if (c >= 100 && c <= 107) s.bg = BRIGHT[c - 100]
      // Anything else is a code the mudlib does not emit; ignoring it is better
      // than rendering the escape as text.
    }

    // Only break the span if the style actually moved. A no-op `ESC[m` in the
    // middle of a word should not fragment it.
    if (!sameStyle(this.style, before)) {
      const after = { ...this.style }
      this.style = before
      this.#closeSpan()
      this.style = after
    }
  }
}

/// `38`/`48` take a sub-form: `;5;N` for indexed, `;2;R;G;B` for 24-bit.
function extended(codes, i) {
  const kind = codes[i + 1]
  if (kind === 5 && codes.length > i + 2) return { color: indexed(codes[i + 2] & 0xff), used: 3 }
  if (kind === 2 && codes.length > i + 4) {
    const [r, g, b] = [codes[i + 2] & 0xff, codes[i + 3] & 0xff, codes[i + 4] & 0xff]
    return { color: `#${hex2(r)}${hex2(g)}${hex2(b)}`, used: 5 }
  }
  return null
}

/// A span's style as inline CSS. `reverse` is done by swapping the two colours
/// rather than with a filter, so it composes with everything else.
export function css(style) {
  const fg = style.reverse ? (style.bg ?? 'var(--bg)') : style.fg
  const bg = style.reverse ? (style.fg ?? 'var(--fg)') : style.bg
  const out = []
  if (fg) out.push(`color:${fg}`)
  if (bg) out.push(`background:${bg}`)
  if (style.bold) out.push('font-weight:700')
  if (style.dim) out.push('opacity:.6')
  if (style.italic) out.push('font-style:italic')
  const lines = []
  if (style.underline) lines.push('underline')
  if (style.strike) lines.push('line-through')
  if (lines.length) out.push(`text-decoration:${lines.join(' ')}`)
  if (style.blink) out.push('animation:ansi-blink 1s steps(2,start) infinite')
  return out.join(';')
}
